toolchains/args.bzl

```diff
@@ -0,0 +1,297 @@
+        allowed_actions: Optional[List[ActionTypeSetInfo]] A list of actions allowed for this action. Should be the intersection of the actions of all variables being accessed.
+        allowed_actions_cause: Optional[str] What to report of the cause of the failure if the actions didn't match.
+        storage: Any metadata to be stored, accessible via self.storage in the fill function.
+
+    Returns:
+        providers suitable for
+    """
+    actions = tuple([action.label for action in collect_action_types(ctx.attr.actions).to_list()])
+    if allowed_actions != None:
+        for action in actions:
+            if action not in allowed_actions:
+                fmt = " ".join(sorted(["  %s\n" % x for x in allowed_actions]))
+                fail("%s is not a valid action for %s, which restricts the actions to:\n%s" % (action, allowed_actions_cause, fmt))
+
+    args = ArgsInfo(
+        label = ctx.label,
+        actions = actions,
+        requires = [target[FeatureConstraintInfo] for target in ctx.attr.requires],
+        files = collect_files(ctx.attr.data),
+        env = ctx.attr.env,
+        fill = fill,
+        storage = storage,
+    )
+    return [
+        args,
+        ArgsListInfo(label = ctx.label, args = tuple([args])),
+    ]
+
+def _args_add_strings_impl(ctx):
+    def fill(self, execution_ctx, args):
+        args.add_all(self.storage)
+        execution_ctx.env.update(self.env)
+
+    return base_add_args_impl(ctx, fill, allowed_actions = None, allowed_actions_cause = None, storage = ctx.attr.values)
+
+args_add_strings = rule(
+    implementation = _args_add_strings_impl,
+    attrs = {
+        "values": attr.string_list(mandatory = True, doc = "The strings to add to the command-line"),
+    } | COMMON_ATTRS,
+)
+
+def get_variable(self, execution_ctx):
+    """Retrieves a variable from the execution context.
+
+    Args:
+        self: (VariableInfo) The variable to retrieve.
+        execution_ctx: (ExecutionContext) The context of execution.
+
+    Returns:
+        The value of the variable in the execution context."""
+    variable = self.storage.variable
+    value = getattr(execution_ctx.variables, variable.name, _MISSING)
+    if value == _MISSING:
+        fail("Variable %s was not provided, but is referenced by %s." % (variable.label, self.label))
+    elif value == None and variable.mandatory:
+        fail("Variable %s was None, but is marked as mandatory" % variable.label)
+    return value
+
+def _args_add_impl(ctx):
+    def fill(self, execution_ctx, args):
+        value = get_variable(self, execution_ctx)
+        if value != None:
+            if self.storage.arg_name != None:
+                args.add(self.storage.arg_name, value, format = self.storage.format)
+            else:
+                args.add(value, format = self.storage.format)
+            execution_ctx.env.update(self.env)
```

Comment:
What do you think of prohibiting collisions instead of potentially relying on ordering?

Reply:
I'm not sure if I agree with it, but since it's much easier to change from disallowed to allowed, I'll just leave it as is for now.